docs/design-proposals/network/dns-of-multi-tenent.md

```diff
@@ -0,0 +1,63 @@
+
+Its full name, in DNS term, should be something like "kube-dns.kube-system.<system-tenant-name>.svc.cluster.local" (system-tenant-name to be decided yet).
+
+## DNS Pod
+Cluster starts a deployment which specifies number of instances of DNS pods. The choices of DNS container image include kubedns, CoreDNS etc. The default DNS image used by Kubernetes v1.11+ is CoreDNS.
+
+With introduction of multi-tenants, type A record of service foo inside of namespace bar of tenant baz, in DNS application should the fully qualified name foo.bar.baz.svc.cluster.local, given cluster base domain is cluster.local. DNS binary needs to be able to consume tenant as part of FQDN.
+
+## Pod DNS configuration
+There are a few options of DNS configuration provided by Kubernetes. No matter what option pod uses, its DNS configuration should have following properties, for the example of pod in namespace bar of tenant baz:
+```text
+search bar.baz.svc.cluster.local baz.svc.cluster.local svc.cluster.local cluster.local 
+options ndots:6
+```
+Kubelet is responsible to inject such configuration to pod when pod is started.
+
+Should we want to limit queries inside its tenant only, the search property could be
+```text
+search bar.baz.svc.cluster.local baz.svc.cluster.local
+```
+This however can not prevent pod from querying services of other tenants if FQDN is used directly.
+## Process of Pod to resolve service IP
+When a pod of namespace bar in tenant baz needs to lookup service foo, 
+1. it uses /etc/resolv.conf to identity the DNS service IP, and derives the fully qualified domain name based (FQDN) on the search property - the first FQDN is foo.baz.bar.svc.cluster.local;
+2. pod sends DNS query request to the DNS service IP, to look up foo.bar.baz.svc.cluster.local; 
+3. request is eventually received by a DNS pod (probably redirect by kube-proxy or other means);
+4. DNS pod locates foo.bar.baz.svc.cluster.local and finds its type A VIP address (or a set of IP addresses for headless service), responds with DNS answer;
+5. the pod receives the DNS answer, and gets the type A address(es).
+
+pod in another namespace in same tenant is able to get foo resolved by query of "foo.bar", at the second DNS attempt. Likewise, pod in other tenant does so by query of "foo.bar.baz" at the third DNS attempt.
+
+## Affected Components
+### DNS binary & image
+[CoreDNS](https://github.com/coredns/coredns.git) is chosen as the first DNS binary to support multi-tenancy, due to its adoption in community.
+We may need to fork CoreDNS repo.
+<br/>Code change is around consuming tenant property from service definition to produce proper FQDN.
+<br/>Other related work includes image building and publishing (assuming we have proper image registry account).
+### DNS pod/service
+DNS pod/service yaml file needs minor change reflecting the new binary image.
```

Comment:
Which IPs will the DNS pods have, so that all customers pods (in various VPCs) have access to it?